### ISSUE TITLE
Thoughts on not using named sources to simplify poetry

### ISSUE THREAD

**Issue by vlcinsky** [sdispater/poetry#1096]
- [x] I have searched the [issues](https://github.com/sdispater/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] I have searched the [documentation](https://poetry.eustace.io/docs/) and believe that my question is not covered.

# Table of Contents

1.  [TL; DR](#org7390ee6)
2.  [Background](#org6a4c956)
    1.  [Separation of concerns](#org7978b71)
    2.  [Tools (incomplete list of examples)](#org7023b04)
        1.  [Local development tools](#org281ce4f)
        2.  [pypi servers](#org83d4c9a)
    3.  [Python package namespace](#orgeaa07e8)
    4.  [Types of software projects: libs and apps](#org310a76b)
3.  [Proposed approach for `poetry`](#orgc359b7f)
4.  [Reasoning](#org1ec5244)
5.  [Dreaming of ideal workflow with packages depending on multiple sources](#orgaeba0ce)

`poetry` and other tools like `pipenv` seem to go in direction of supporting named pypi indexes (called **sources** by `poetry` when in role of index page providing python packages). During developing feature proposal to make `poetry` easier to use during CI/CD we have realised, that this feature adds quite some complexity in many directions while removing such feature would contribute to simplification and ease of `poetry` use.

This issue is here to discuss such option.

<a id="org7390ee6"></a>

# TL; DR

Managing multiple package sources by `poetry` has some disadvantages:

-   it adds extra complexity [1]
-   created python package (e.g. as wheel) forgets [2] sources as declared dependencies within package know only package names
-   even application installed by means of `poetry` gets more shaky, if it depends on more sources

better:

-   use private pypi (or even wheelhouse), which takes care of merging packages from multiple sources
-   let `poetry` install from single (private) or two (private and public) package sources

Notes:

-   [1] complexity added to dependencies in `pyproject.toml` and `poetry.lock`, to `poetry` global config and to style `poetry` can be used e.g. in CI/CD jobs
-   [2] there seem to be some url options in [PEP-508](<https://www.python.org/dev/peps/pep-0508/>), but I would not dare using it in real project

<a id="org6a4c956"></a>

# Background

Before going to specific conclusions, few contextual information is presented.

<a id="org7978b71"></a>

## Separation of concerns

Tools may deal with one or more concerns such as: 

-   provide packages for installation (pypi indexes)
-   provide packages from multiple indexes
-   define dependencies for an application
-   define dependencies for python package
-   package source code into python package
-   install python package incl. it's dependencies
-   create virtual env

Tools typically cover more concerns at once, but trying to cover too many might create a problem.

<a id="org7023b04"></a>

## Tools (incomplete list of examples)

Here are some tools relevant to this discussion (definitely not exhaustive list, just good samples):

<a id="org281ce4f"></a>

### Local development tools

These tools typical deal with managing venvs, installing packages or supporting package creation:

-   `venv`: virtual environment creation tool
-   `pip`: python package installer
-   `pipenv`:
    -   virtual env management tool
    -   definitely not "packaging tool" in terms of python (library) package
-   `flit`:
    -   packaging tool
-   `poetry`:
    -   python package creation tool (real packaging tool)
    -   virtual env management tool

<a id="org83d4c9a"></a>

### pypi servers

Pypi servers serve as source of python packages. They must allow to install published packages from them and possibly allow publishing packages to them.

-   <https://pypi.org>: public pypi server
-   [`pypicloud`](https://pypicloud.readthedocs.io/en/latest/): pypi server.
    -   great for installation in production from single logical index
    -   can be secured
    -   can scale up well
    -   limited support of some features needed during development (e.g. missing separate user indexes)
-   `devpi-server`: private pypi index
    -   serving well multiple indexes, multiple users
    -   allows to integrate packages from multiple sources (any index can define set of "bases" which are searched, when previous one did not provide the package)
    -   great for in-house use
    -   not intended to run in "wild internet" (securing it is not easy).
    -   `devpi`: devpi-server client (to manage python package on `devpi-server`). Note: install it from `devpi-client` package, `devpi` one is obsolete.
-   [`gemfury`](https://gemfury.com/l/pypi-server): pypi server as SaaS (I am not familiar with full feature set)

<a id="orgeaa07e8"></a>

## Python package namespace

Python package namespace is in general one global one.

While there are some options to declare, that given python package is to be installed from specific source, such option is not supported everywhere, and it is often optional feature.

<a id="org310a76b"></a>

## Types of software projects: libs and apps

-   applications:
    -   shall have very deterministic way of installing, what requires not only freezing versions, but even using package hashes.
-   libraries (python package):
    -   python package must declare dependencies more flexibly. Restrictions shall be based on availability of expected features in specific versions, but being too strict may prevent installation of the package together with other packages being dependent on the same external library in a bit different version.
    -   library has no or very rare means to define multiple package sources (it only declares dependencies)

<a id="orgc359b7f"></a>

# Proposed approach for `poetry`

-   for uploading to repositories, keep it as it is now. This is out of scope of this proposal.
-   for installing packages from multiple sources, take one of approaches:
    -   [reject-sources] reject such option completely and rely on ability of the developer/deployer to provide single source with all packages needed in it.
    -   [index-url] allow such option, but make it easy to change source urls by single option on CLI
        -   \`index-url\` (in style of `$ pip install --index-url` option)
    -   [index-extra-url] allow such option, but make it easy to change source urls by two options on CLI:
        -   \`index-url\` (in style of `$ pip install --index-url` option, but allow only one such instance)
        -   \`extra-url\` (in style of `$ pip install --extra-url` option, but allow only one such instance)
-   additionally, few working examples of managing multiple sources via other tools such as \`devpi-server\` would help users (if there is such interest, I can try to write one)
-   additionally it would be great, if \`index-url\` (and possibly \`extra-url\`) allow installing from a directory (such as wheelhouse).

<a id="org1ec5244"></a>

# Reasoning

Managing dependencies from multiple sources is concern, which is better managed out of packaging tool. Such tools already exists (e.g. \`devpi-server\`, download directories etc.) and they seem to be much more flexible in situation like testing, deployment, installing without access to external pypi indexes etc.

Problems with using multiple sources in dependency declaration are these:

-   \`poetry\` usage becomes more complex (not only when starting with such option, but later supporting this in other sub-commands and scenarios becomes difficult)
-   built package theoretically allows reference to package source (see PEP508), but such feature is not yet widely used and using such package might be brittle
    -   problems to access (multiple) external indexes might prevent or break installation
    -   it effectively prevents testing against test and production (or even more stages) of dependent package builds.
-   CI/CD becomes more difficult.
    -   If any of sources becomes not available, build cannot run
    -   if one plans for builds, which do not depend on external sources, it is not clear how to achieve such goal
    -   using test, staging and production index variants becomes impossible as index sources must be fixed

<a id="orgaeba0ce"></a>

# Dreaming of ideal workflow with packages depending on multiple sources

This dream assumes one is building a python package or application for private usage and plan to run automated CI/CD with it.

-   kick off development
    -   identify packages from extra sources
    -   make sure, devpi-server is installed (this is question of minutes, see tutorial)
    -   in devpi-server:
        -   create or select and index to use for this project
        -   download required packages and upload them to this index
    -   configure poetry to use given index and specify it via \`&#x2013;index-url\` option, configuration option of env variable
    -   work on the poetry based project
    -   finally upload the resulting package to given index
-   create environment for staging
    -   using devpi-server, create new index
    -   push there all the packages needed (now use \`devpi push\` and mostly push tested packages from previous index to the new one)
    -   build CI/CD based on such index
    -   when things need to be fixed, update packages in this staging index
    -   finally, create \`download\` directory where all the packages needed are used
-   creating environment for deployment
    -   if you plan to install from an index, create one and upload there all packages needed (created in previous stage)
    -   install from this index
    -   alternatively, deploy directly from the "download" directory (using e.g. proper requirement.txt)

**Comment by vlcinsky** [comment]
This issue assumes, that @sdispater plans to use named sources as optional argument when declaring dependencies. See Background section in #1070 where are references to drafted code.
